lib/ide/parser/src/api.rs

```diff
@@ -1,20 +1,57 @@
 use crate::prelude::*;
+use ast::ID;
+use std::ops::Sub;
 
-pub type Ast = ast::Ast;
+use serde::Serialize;
+use serde::Deserialize;
 
+pub type Ast = ast::Ast;
 
 
 // ============
 // == Parser ==
 // ============
 
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Index { pub value:usize }
```

Comment:
I'd move `Index`, `Size` and `Span` to `utilities` crate (e.g. to some path like `src/data/text.rs`.
These are useful dictionary type that I'd encourage various part of codebase to use (and add more utilities to them). I don't consider them specific to parser. (particularly since we plan to use them in ast)